### PR TITLE
RavenDB-16345 Start index when enabling index cluster wide

### PR DESCRIPTION
**Pull request by efratshenhar**

### CODE REVIEW

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 1]
```diff
@@ -903,14 +903,27 @@ public virtual void Update(IndexDefinitionBase definition, IndexingConfiguration
...
+                    if (definition.State == IndexState.Normal)
+                    {
+                        SetState(IndexState.Normal);
+                        stateChanged = true;
+                    }
+                    else
+                    {
+                        SetState(definition.State);
+                    }
+                }
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/Index.cs]: Can it be simplified to :
```
stateChanged = definition.State == IndexState.Normal;
SetState(definition.State);
```

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 2]
```diff
@@ -903,14 +903,27 @@ public virtual void Update(IndexDefinitionBase definition, IndexingConfiguration
...
+
                 Definition = definition;
                 Configuration = configuration;
 
                 OnInitialization();
 
                 _priorityChanged.Raise();
 
-                if (status == IndexRunningStatus.Running)
+                if ((status == IndexRunningStatus.Running) || ((stateChanged) && (definition.State == IndexState.Normal)))
```
Comment (karmeli87) [inline src/Raven.Server/Documents/Indexes/Index.cs]: why `((status == IndexRunningStatus.Running) || stateChanged)` isn't enough here?

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 1 above)
Comment (karmeli87) [reply src/Raven.Server/Documents/Indexes/Index.cs]: Does `stateChanged` means `EnableRequired` ?

File: src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+        /// Priority of an index
+        /// </summary>
+        public IndexPriority? Priority { get; set; }
+
+        /// <summary>
+        /// Index state
+        /// </summary>
+        public IndexState? State { get; set; }
+
+        public long LastStateChangeRaftIndex { get; set; }
```
Comment (karmeli87) [inline src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs]: Need to think, how to hide this. Making it `internal` mess up the de-serialization. 
cc @ppekrol

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 1 above)
Comment (efratshenhar) [reply src/Raven.Server/Documents/Indexes/Index.cs]: yes

File: src/Raven.Server/Documents/Indexes/Index.cs (hunk 2 above)
Comment (efratshenhar) [reply src/Raven.Server/Documents/Indexes/Index.cs]: the state can be changed to disabled

Comment (karmeli87) [reply src/Raven.Server/Documents/Indexes/Index.cs]: but we set it to true only if `definition.State == IndexState.Normal`, or do I miss something?

Comment (efratshenhar) [reply src/Raven.Server/Documents/Indexes/Index.cs]: yes, you are right

File: src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs [hunk 4]
```diff
@@ -0,0 +1,26 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Raven.Client.Documents.Indexes
+{
+    public abstract class ClientIndexDefinitionBase
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs]: Let's rename to IndexDefinitionBase

File: src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs (hunk 3 above)
Comment (ppekrol) [reply src/Raven.Client/Documents/Indexes/ClientIndexDefinitionBase.cs]: Can you show me that?

File: src/Raven.Server/Documents/Indexes/IndexDefinitionBase.cs [hunk 5]
```diff
@@ -29,6 +29,8 @@ public abstract class IndexDefinitionBase
 
         public IndexState State { get; set; }
 
+        internal long LastStateChangeRaftIndex { get; set; }
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/IndexDefinitionBase.cs]: rename this class to IndexDefinitionBaseServerSide. We already have similar convention with other classes like that.

File: src/Raven.Client/Documents/Indexes/IndexDefinition.cs [hunk 6]
```diff
@@ -15,28 +15,14 @@ namespace Raven.Client.Documents.Indexes
     /// <summary>
     /// A definition of a RavenIndex
     /// </summary>
-    public class IndexDefinition
+    public class IndexDefinition : IndexDefinitionBase
     {
         public IndexDefinition()
         {
             _configuration = new IndexConfiguration();
+            ClusterIndex = new ClusterIndex();
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/IndexDefinition.cs]: Shouldn't this be in ctor of IndexDefinitionBase?

File: src/Raven.Client/Documents/Indexes/ClusterIndex.cs [hunk 7]
```diff
@@ -0,0 +1,7 @@
+﻿namespace Raven.Client.Documents.Indexes
+{
+    public class ClusterIndex
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/ClusterIndex.cs]: ClusterState?

File: src/Raven.Client/Documents/Indexes/ClusterIndex.cs [hunk 8]
```diff
@@ -0,0 +1,7 @@
...
+    {
+        public long? ClusterIndexForState;
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/ClusterIndex.cs]: LastStateIndex?

File: src/Raven.Client/Documents/Indexes/IndexDefinition.cs [hunk 9]
```diff
@@ -469,7 +455,8 @@ internal IndexDefinition Clone()
                 OutputReduceToCollection = OutputReduceToCollection,
                 ReduceOutputIndex = ReduceOutputIndex,
                 PatternForOutputReduceToCollectionReferences = PatternForOutputReduceToCollectionReferences,
-                PatternReferencesCollectionName = PatternReferencesCollectionName
+                PatternReferencesCollectionName = PatternReferencesCollectionName,
+                ClusterIndex = new ClusterIndex(){ClusterIndexForState = ClusterIndex?.ClusterIndexForState ?? 0}
```
Comment (ppekrol) [inline src/Raven.Client/Documents/Indexes/IndexDefinition.cs]: create a copy ctor

Comment (ppekrol) [reply src/Raven.Client/Documents/Indexes/IndexDefinition.cs]: also this does not make a 1:1 copy, is this is what we want?

File: src/Raven.Client/Json/Serialization/NewtonsoftJson/DefaultRavenContractResolver.cs [hunk 10]
```diff
@@ -193,7 +193,7 @@ private bool MembersToFilterOut(MemberInfo info)
             var fieldInfo = info as FieldInfo;
             if (fieldInfo != null)
             {
-                if (fieldInfo.IsPublic == false)
+                if ((fieldInfo.IsPublic == false) && (fieldInfo.IsDefined(typeof(JsonDeserializationDoNotIgnoreAttribute)) == false))
```
Comment (ppekrol) [inline src/Raven.Client/Json/Serialization/NewtonsoftJson/DefaultRavenContractResolver.cs]: What about property? Handled 20 lines below.

File: src/Raven.Server/Documents/Indexes/Index.cs [hunk 11]
```diff
@@ -911,7 +917,7 @@ public virtual void Update(IndexDefinitionBase definition, IndexingConfiguration
 
                 _priorityChanged.Raise();
 
-                if (status == IndexRunningStatus.Running)
+                if ((status == IndexRunningStatus.Running) || (stateChanged))
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/Index.cs]: redundant parenthesis ?

File: src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs [hunk 12]
```diff
@@ -171,9 +175,10 @@ protected IndexDefinitionBase(string name, IEnumerable<string> collections, Inde
             Priority = priority;
             State = state;
             _indexVersion = indexVersion;
+            ClusterIndex.ClusterIndexForState = clusterIndexForState ?? 0;
```
Comment (ppekrol) [inline src/Raven.Server/Documents/Indexes/IndexDefinitionBaseServerSide.cs]: I see that 0 is some default value here? Noticed it in few places. If yes then maybe this field should not be nullable?

File: src/Raven.Server/Extensions/JsonSerializationExtensions.cs [hunk 13]
```diff
@@ -24,6 +24,10 @@ public static DynamicJsonValue ToJson(this IndexDefinition definition)
             result[nameof(IndexDefinition.Reduce)] = definition.Reduce;
             result[nameof(IndexDefinition.Type)] = definition.Type.ToString();
             result[nameof(IndexDefinition.Maps)] = new DynamicJsonArray(definition.Maps);
+            result[nameof(IndexDefinition.ClusterIndex)] = new DynamicJsonValue()
+            {
+                [nameof(IndexDefinition.ClusterIndex.ClusterIndexForState)] = definition.ClusterIndex?.ClusterIndexForState ?? 0
```
Comment (ppekrol) [inline src/Raven.Server/Extensions/JsonSerializationExtensions.cs]: 0 here again

File: src/Sparrow/Json/JsonDeserializationBase.cs [hunk 14]
```diff
@@ -66,8 +66,11 @@ public class JsonDeserializationBase
                 }
 
                 var propInit = new List<MemberBinding>();
-                foreach (var fieldInfo in typeof(T).GetFields())
+                foreach (var fieldInfo in typeof(T).GetFields(BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance))
                 {
+                    if ((fieldInfo.IsPublic == false) && (fieldInfo.IsDefined(typeof(JsonDeserializationDoNotIgnoreAttribute)) == false))
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonDeserializationBase.cs]: Handle properties also?

File: src/Sparrow/Json/JsonDeserializationDoNotIgnoreAttribute.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
+
+namespace Sparrow.Json
+{
+    /// <summary>
+    /// Instructs the <see cref="JsonDeserializationBase"/> to serialize nonPublic field .
+    /// </summary>
+    [AttributeUsage(AttributeTargets.Field )]
+    public sealed class JsonDeserializationDoNotIgnoreAttribute : Attribute
```
Comment (ppekrol) [inline src/Sparrow/Json/JsonDeserializationDoNotIgnoreAttribute.cs]: internal

Comment (ppekrol) [inline src/Sparrow/Json/JsonDeserializationDoNotIgnoreAttribute.cs]: JsonDeserializationNoIgnoreAttribute maybe?

File: src/Raven.Client/Json/Serialization/NewtonsoftJson/DefaultRavenContractResolver.cs (hunk 10 above)
Comment (efratshenhar) [reply src/Raven.Client/Json/Serialization/NewtonsoftJson/DefaultRavenContractResolver.cs]: no need. we are not filtering out non-public properties

**Pull request closed without merging**